File: check_dom.py
import re
import os
import sys

try:
    current_dir = os.path.dirname(os.path.abspath(__file__))
    js_path = os.path.join(current_dir, 'app.js')
    html_path = os.path.join(current_dir, 'index.html')
    
    with open(js_path, "r", encoding="utf-8") as f:
        js_content = f.read()

    with open(html_path, "r", encoding="utf-8") as f:
        html_content = f.read()

    # Extract all document.getElementById("...") or '...'
    js_ids = set(re.findall(r'document\.getElementById\([\'"]([a-zA-Z0-9_-]+)[\'"]\)', js_content))

    # Also look for querySelector or check all IDs in HTML
    html_ids = set(re.findall(r'id=[\'"]([a-zA-Z0-9_-]+)[\'"]', html_content))

    output = []
    output.append(f"Total IDs found in JS: {len(js_ids)}")
    output.append(f"Total IDs found in HTML: {len(html_ids)}")

    missing_in_html = js_ids - html_ids
    if missing_in_html:
        output.append(f"CRITICAL WARNING: These IDs are in JS but MISSING in HTML: {missing_in_html}")
    else:
        output.append("SUCCESS: All JS DOM element IDs exist in HTML!")
        
    result_str = "\n".join(output)
    print(result_str)
    
    output_path = os.path.join(current_dir, 'output.txt')
    with open(output_path, "w", encoding="utf-8") as out_f:
        out_f.write(result_str)
except Exception as e:
    print(f"ERROR: {str(e)}")
    sys.exit(1)

File: check_dom.js
const fs = require('fs');
const path = require('path');

try {
    const jsContent = fs.readFileSync(path.join(__dirname, 'app.js'), 'utf8');
    const htmlContent = fs.readFileSync(path.join(__dirname, 'index.html'), 'utf8');

    // Regex for document.getElementById("...") or '...'
    const jsIds = new Set();
    const idRegex = /document\.getElementById\(['"]([a-zA-Z0-9_-]+)['"]\)/g;
    let match;
    while ((match = idRegex.exec(jsContent)) !== null) {
        jsIds.add(match[1]);
    }

    // Regex for id="..." or '...' in HTML
    const htmlIds = new Set();
    const htmlIdRegex = /id=['"]([a-zA-Z0-9_-]+)['"]/g;
    while ((match = htmlIdRegex.exec(htmlContent)) !== null) {
        htmlIds.add(match[1]);
    }

    const missingInHtml = [...jsIds].filter(id => !htmlIds.has(id));

    const output = [];
    output.push(`Total IDs found in JS: ${jsIds.size}`);
    output.push(`Total IDs found in HTML: ${htmlIds.size}`);

    if (missingInHtml.length > 0) {
        output.push(`CRITICAL WARNING: These IDs are in JS but MISSING in HTML: ${missingInHtml.join(', ')}`);
    } else {
        output.push("SUCCESS: All JS DOM element IDs exist in HTML!");
    }

    const resultStr = output.join('\n');
    console.log(resultStr);

    fs.writeFileSync(path.join(__dirname, 'output.txt'), resultStr, 'utf8');
} catch (e) {
    fs.writeFileSync(path.join(__dirname, 'output.txt'), `ERROR: ${e.message}`, 'utf8');
}

